graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java

```diff
@@ -171,6 +194,72 @@ public UserList listUsers() {
+    public UserPageListResponse getPage(@ApiParam(name = "page") @QueryParam("page") @DefaultValue("1") int page,
+                                        @ApiParam(name = "per_page") @QueryParam("per_page") @DefaultValue("50") int perPage,
+                                        @ApiParam(name = "query") @QueryParam("query") @DefaultValue("") String query,
+                                        @ApiParam(name = "sort",
+                                                value = "The field to sort the result on",
+                                                required = true,
+                                                allowableValues = "title,description")
+                                            @DefaultValue(StreamImpl.FIELD_TITLE) @QueryParam("sort") String sort,
+                                        @ApiParam(name = "order", value = "The sort direction", allowableValues = "asc, desc")
+                                            @DefaultValue("asc") @QueryParam("order") String order) {
+
+        SearchQuery searchQuery;
+        final Collection<MongoDbSession> sessions = sessionService.loadAll();
+        final Map<String, Optional<MongoDbSession>> lastSessionForUser = getLastSessionForUser(sessions);
+        try {
+            searchQuery = searchQueryParser.parse(query);
+        } catch (IllegalArgumentException e) {
+            throw new BadRequestException("Invalid argument in search query: " + e.getMessage());
+        }
+
+        final PaginatedList<UserDTO> result = paginatedUserService
+                .findPaginated(searchQuery, page, perPage, sort, order);
+        final Set<String> allRoleIds = result.stream().flatMap(userDTO -> {
+            if (Optional.ofNullable(userDTO.roles()).isPresent()) {
+                return userDTO.roles().stream();
+            }
+            return Stream.empty();
+        }).collect(Collectors.toSet());
+
+        Map<String, String> roleNameMap;
+        try {
+            roleNameMap = getRoleNameMap(allRoleIds);
+        } catch (org.graylog2.database.NotFoundException e) {
+            throw new NotFoundException("Couldn't find roles: " + e.getMessage());
+        }
+
+        final UserDTO adminUser = getAdminUserDTO(lastSessionForUser, roleNameMap);
+
+        List<UserDTO> users = result.stream().map(userDTO -> {
+            UserDTO.Builder builder = userDTO.toBuilder()
+                    .fillSession(lastSessionForUser.getOrDefault(userDTO.username(), Optional.empty()));
+            if (Optional.ofNullable(userDTO.roles()).isPresent()) {
+                builder.roles(userDTO.roles().stream().map(roleNameMap::get).collect(Collectors.toSet()));
+            }
+            return builder.build();
+        }).collect(Collectors.toList());
+
+        final PaginatedList<UserDTO> userDTOS = new PaginatedList<>(users, result.pagination().total(),
+                result.pagination().page(), result.pagination().perPage());
+        return UserPageListResponse.builder()
```

Comment:
I think we can use the `PaginatedResponse` return value here. That would make the response similar to what we return from other paginated endpoints. (with the custom `UserPageListResponse` we have a `pagination` key inside the response, `PaginatedResponse` doesn't have that) The `admin_user` will move to under the `context` key in the response, but I think that's okay.

```json
return PaginatedResponse.create("users", userDTOS, query, Collections.singletonMap("admin_user", adminUser));
```